src/sparse/in_memory.rs

```diff
@@ -0,0 +1,137 @@
+use crate::common::Bytes32;
+use crate::sparse::Buffer;
+use crate::{common, sparse};
+
+use alloc::boxed::Box;
+use core::marker::PhantomPinned;
+use core::pin::Pin;
+use core::ptr::NonNull;
+
+type StorageMap = common::StorageMap<Bytes32, Buffer>;
+type SparseMerkleTree<'a> = sparse::MerkleTree<'a, StorageMap>;
+
+pub struct MerkleTree<'a> {
+    storage: StorageMap,
+    tree: Option<SparseMerkleTree<'a>>,
+    _marker: PhantomPinned,
+}
+
+impl<'a> MerkleTree<'a> {
+    pub fn new() -> Pin<Box<Self>> {
```

Comment:
Because this is a self-referential structure, I use `Pin`. This means that the rest of the interface has signatures that have `self: Pin<&mut Self>`, and callers will look something like `tree.as_mut().update(...)`.

From the perspective of a user of this crate, will this work with our other libs?

Reply:
Check if this simplifies a thing or two: https://github.com/FuelLabs/fuel-merkle/pull/101/files